File: Assets/Scripts/Networking/Protocol/WorldEntityRemovedCodec.cs
using System;

namespace TopDownRoguelike.Networking.Protocol
{
    public enum WorldEntityRemovalReason : byte
    {
        Invalid = 0,
        Died = 1,
        Cleared = 2,
        Despawned = 3
    }

    public sealed class WorldEntityRemovedPayload
    {
        public WorldEntityRemovedPayload(
            uint entityId,
            NetworkEntityType entityType,
            WorldEntityRemovalReason reason)
        {
            EntityId = entityId;
            EntityType = entityType;
            Reason = reason;
        }

        public uint EntityId { get; }
        public NetworkEntityType EntityType { get; }
        public WorldEntityRemovalReason Reason { get; }
    }

    public static class WorldEntityRemovedCodec
    {
        public const int PayloadSize = 8;

        public static byte[] Encode(
            WorldEntityRemovedPayload removed)
        {
            Validate(removed);

            var payload = new byte[PayloadSize];

            PacketCodec.WriteNetworkUInt32(
                payload,
                0,
                removed.EntityId);

            payload[4] = (byte)removed.EntityType;
            payload[5] = (byte)removed.Reason;

            return payload;
        }

        public static WorldEntityRemovedPayload Decode(
            byte[] payload)
        {
            if (payload == null)
            {
                throw new ArgumentNullException(nameof(payload));
            }

            if (payload.Length != PayloadSize)
            {
                throw new ArgumentException(
                    "World entity removal payload must be 8 bytes.",
                    nameof(payload));
            }

            if (payload[6] != 0 || payload[7] != 0)
            {
                throw new ArgumentException(
                    "World entity removal reserved bytes must be zero.",
                    nameof(payload));
            }

            var removed =
                new WorldEntityRemovedPayload(
                    PacketCodec.ReadNetworkUInt32(payload, 0),
                    (NetworkEntityType)payload[4],
                    (WorldEntityRemovalReason)payload[5]);

            Validate(removed);
            return removed;
        }

        private static void Validate(
            WorldEntityRemovedPayload removed)
        {
            if (removed == null)
            {
                throw new ArgumentNullException(nameof(removed));
            }

            if (removed.EntityId == 0u)
            {
                throw new ArgumentException(
                    "Removed entity ID must be non-zero.",
                    nameof(removed));
            }

            int entityType = (int)removed.EntityType;

            if (entityType < (int)NetworkEntityType.Player ||
                entityType > (int)NetworkEntityType.ExperienceOrb)
            {
                throw new ArgumentException(
                    "Removed entity type is invalid.",
                    nameof(removed));
            }

            int reason = (int)removed.Reason;

            if (reason < (int)WorldEntityRemovalReason.Died ||
                reason > (int)WorldEntityRemovalReason.Despawned)
            {
                throw new ArgumentException(
                    "Entity removal reason is invalid.",
                    nameof(removed));
            }
        }
    }
}

File: NetworkServer/src/protocol/WorldEntityRemovedCodec.cpp
#include "protocol/WorldEntityRemovedCodec.h"

#include <stdexcept>

namespace
{
    void Validate(
        const tdr::protocol::WorldEntityRemovedPayload& payload
    )
    {
        using namespace tdr::protocol;

        const auto entityType =
            static_cast<std::uint8_t>(payload.entityType);
        const auto reason =
            static_cast<std::uint8_t>(payload.reason);

        if (payload.entityId == 0U)
        {
            throw std::invalid_argument(
                "Removed entity ID must be non-zero."
            );
        }

        if (entityType < static_cast<std::uint8_t>(
                WorldEntityType::Player) ||
            entityType > static_cast<std::uint8_t>(
                WorldEntityType::BossProjectile))
        {
            throw std::invalid_argument(
                "Removed entity type is invalid."
            );
        }

        if (reason < static_cast<std::uint8_t>(
                WorldEntityRemovalReason::Died) ||
            reason > static_cast<std::uint8_t>(
                WorldEntityRemovalReason::Despawned))
        {
            throw std::invalid_argument(
                "Entity removal reason is invalid."
            );
        }
    }
}

namespace tdr::protocol
{
    std::vector<std::uint8_t>
        WorldEntityRemovedCodec::Encode(
            const WorldEntityRemovedPayload& payload
        )
    {
        Validate(payload);

        return {
            static_cast<std::uint8_t>(payload.entityId >> 24U),
            static_cast<std::uint8_t>(payload.entityId >> 16U),
            static_cast<std::uint8_t>(payload.entityId >> 8U),
            static_cast<std::uint8_t>(payload.entityId),
            static_cast<std::uint8_t>(payload.entityType),
            static_cast<std::uint8_t>(payload.reason),
            0U,
            0U
        };
    }

    WorldEntityRemovedPayload
        WorldEntityRemovedCodec::Decode(
            const std::uint8_t* const data,
            const std::size_t size
        )
    {
        if (data == nullptr)
        {
            throw std::invalid_argument(
                "World entity removal payload is null."
            );
        }

        if (size != kWorldEntityRemovedPayloadSize)
        {
            throw std::invalid_argument(
                "World entity removal payload must be 8 bytes."
            );
        }

        if (data[6] != 0U || data[7] != 0U)
        {
            throw std::invalid_argument(
                "World entity removal reserved bytes must be zero."
            );
        }

        WorldEntityRemovedPayload payload{};
        payload.entityId =
            (static_cast<std::uint32_t>(data[0]) << 24U) |
            (static_cast<std::uint32_t>(data[1]) << 16U) |
            (static_cast<std::uint32_t>(data[2]) << 8U) |
            static_cast<std::uint32_t>(data[3]);
        payload.entityType =
            static_cast<WorldEntityType>(data[4]);
        payload.reason =
            static_cast<WorldEntityRemovalReason>(data[5]);

        Validate(payload);
        return payload;
    }
}
